### PR TITLE
test: tendermint client update when client expires or validator set has changed

### LINKED ISSUE #538: [ICS02] Cover client frozen/expired conditions with unit tests
## Problem Statement

Make sure that we covered failure scenarios everywhere we check a status of a client (like [here](https://github.com/cosmos/ibc-rs/blob/45d5f0d25f323eb423d5511e2488be8104e1c39f/crates/ibc/src/core/ics02_client/handler/misbehaviour.rs#L27-L29)) by introducing frozen and expired conditions. 
Adding required fields to the `MockCotext` and writing unit tests should accomplish this.

Should come after fixing https://github.com/cosmos/ibc-rs/issues/536

### PR DESCRIPTION
**Pull request by rnbguy**
Closes: #538 

## Description

This PR adds and modifies a few test scenarios.

1. Client expiry while client state update.
2. Client state update with validator set changes (happy and sad).

______

### PR author checklist:
- [x] Added changelog entry, using [`unclog`](https://github.com/informalsystems/unclog).
- [ ] Create a new issue for parameterized mock Header, Client, and Context configuration.
- [x] Added tests.
- [x] Linked to GitHub issue.
- [x] Updated code comments and documentation (e.g., `docs/`).
- [x] Tagged *one* reviewer who will be the one responsible for shepherding this PR.

### Reviewer checklist:

- [ ] Reviewed `Files changed` in the GitHub PR explorer.
- [ ] Manually tested (in case integration/unit/mock tests are absent).

### CODE REVIEW

**Comment by rnbguy** [comment]
I am not sure what is going on. The earlier commits failed on gh action, although they were successful on my machine (`cargo test --all-features --no-run` and `act`). I added 0f4503c as a fix - even though with it, the build fails on my machine - but it makes gh actions successful!

@Farhad-Shabani can you please take a look later?

**Comment by rnbguy** [comment]
Did I hit a rare issue with Git?

This is from this PR
https://github.com/cosmos/ibc-rs/blob/0f4503ce85e042efaf89063989d81afe01d23706/crates/ibc/src/mock/context.rs#L240

This is from `main` branch
https://github.com/cosmos/ibc-rs/blob/245d2fc3d53d7b3c9eb6d2380580c798aa150fe9/crates/ibc/src/mock/context.rs#L203

Although it's shown when two commits are [compared](https://github.com/cosmos/ibc-rs/compare/245d2fc3d53d7b3c9eb6d2380580c798aa150fe9..0f4503ce85e042efaf89063989d81afe01d23706).

**Comment by rnbguy** [comment]
~Please ignore my previous comments.~ I just realized the `main` branch progressed from what I had.

The new changes did not conflict with my PR syntactically but they [conflicted semantically](https://bors.tech/essay/2017/02/02/pitch). :sweat_smile:

File: crates/ibc/src/core/ics02_client/handler/update_client.rs
Comment (Farhad-Shabani) [inline crates/ibc/src/core/ics02_client/handler/update_client.rs]: On a general note on such TODO’s annotation, let's keep the comments as clean as we can and instead if there is something we should really get to it later, let’s open an issue (not necessarily a detailed one). This way can better track our tech debts as well. Thanks!

File: crates/ibc/src/mock/context.rs
Comment (Farhad-Shabani) [inline crates/ibc/src/mock/context.rs]: Same comment on `TODO`s applies here as well.

File: crates/ibc/src/mock/host.rs
Comment (Farhad-Shabani) [inline crates/ibc/src/mock/host.rs]: Same comment on `TODO`s applies here as well.

File: crates/ibc/src/mock/host.rs
Comment (Farhad-Shabani) [inline crates/ibc/src/mock/host.rs]: Same comment on `TODO`s applies here as well.

File: crates/ibc/src/mock/host.rs
Comment (Farhad-Shabani) [inline crates/ibc/src/mock/host.rs:42]: Do we have to add this `trusted_next_validators` as a new field here? I mean, we can already get to the same stuff through the `light_block` field. Right?

File: crates/ibc/src/core/ics02_client/handler/update_client.rs
Comment (Farhad-Shabani) [inline crates/ibc/src/core/ics02_client/handler/update_client.rs:257]: `ctx_a` instead of `ctx` ?

File: crates/ibc/src/mock/context.rs
Comment (Farhad-Shabani) [inline crates/ibc/src/mock/context.rs]: Should panic here instead of `todo!()` ?

File: crates/ibc/src/mock/context/clients.rs
Comment (Farhad-Shabani) [inline crates/ibc/src/mock/context/clients.rs]: Here, we should host the code related to the `mock` and implementing contexts on it. 
`TmClientStateConfig` should be placed under the `ics07_tendermint` and enabled using the `mocks` feature flag (along with others (https://github.com/cosmos/ibc-rs/blob/245d2fc3d53d7b3c9eb6d2380580c798aa150fe9/crates/ibc/src/clients/ics07_tendermint/client_state.rs#L1122-L1183))

File: crates/ibc/src/core/ics02_client/handler/update_client.rs
Comment (Farhad-Shabani) [inline crates/ibc/src/core/ics02_client/handler/update_client.rs]: We can remove `.host_type()` and `.max_history_size()` given that we have access to these through default derivations.

File: crates/ibc/src/core/ics02_client/handler/update_client.rs
Comment (Farhad-Shabani) [inline crates/ibc/src/core/ics02_client/handler/update_client.rs]: To be consistent with other places, could we keep using [our naming convention](https://github.com/cosmos/ibc-rs/blob/dc8e4161dd5e00a2031cf3feb58dad1ed7e26bdb/crates/ibc/src/core/mod.rs#L3-L23) instead of remote and host?

File: crates/ibc/Cargo.toml
Comment (Farhad-Shabani) [inline crates/ibc/Cargo.toml]: Should update this to 0.17.0?

File: crates/ibc/src/mock/context.rs
Comment (Farhad-Shabani) [inline crates/ibc/src/mock/context.rs]: Here, we logically should have a single method to advance the host chain, managing both the height and timestamp. This method should not accept the timestamp as an argument to align with the `block_time` field of the `MockContext`. When necessary, advancing by a specific duration it either should be done by mulitple calls or setting the `block_time`` other that its default value.

File: crates/ibc/src/mock/context.rs
Comment (Farhad-Shabani) [inline crates/ibc/src/mock/context.rs:222]: As the entire MockContext is subject to change soon, I didn't delve deeply into abstraction using `TypedBuilder` in this PR. Seeing as it works for our tests, I think it's fine. Though a question I got: couldn’t we directly derive `TypedBuilder` on the own `MockContext`?

**Review by Farhad-Shabani — CHANGES_REQUESTED** [review]

File: crates/ibc/src/mock/host.rs
Comment (rnbguy) [reply crates/ibc/src/mock/host.rs:42]: Very good question. To answer your question, no. (Light-)Blocks do not store `trusted_height` and `trusted_next_validators`. They are computed by the relayer by querying the remote blockchain. Actually, I will suggest removing these two fields from `SyntheticTmBlock` as these two fields are not required for a host context. These fields are only needed when the relayer submits a client update message to the counterparty chain.

Previously, in our Mock version, `trusted_next_validators` was computed as the next validator of the current header - which is semantically wrong. But works for consecutive client height updates.

And the same was being done in the issue reported by Nomic. This is why my previously written test should not have failed. But I made a mistake :sweat_smile: The fix is in a2163f1cd2b4f848ae285ec26bb9f0ac0719d7b1. With this fix, the bug couldn't have been caught by tests.

Note, I can't set `light_block.header.next_validator` later - as this will break the header signature verification.

File: crates/ibc/src/mock/context/clients.rs
Comment (rnbguy) [reply crates/ibc/src/mock/context/clients.rs]: Ah yes. You are right :+1:

File: crates/ibc/src/mock/context.rs
Comment (rnbguy) [reply crates/ibc/src/mock/context.rs:222]: Are we ok with maintaining the macro code for deriving Builders? I don't think we should manually write these setter methods.

File: crates/ibc/src/mock/context.rs
Comment (rnbguy) [reply crates/ibc/src/mock/context.rs]: Hmm. Correct me if I am wrong. It can happen in real life that the chain was halted but later restarted again from a halted height. In that particular height, the block time can be arbitrarily long. There is no block production logic to avoid this. I tried to mimic that situation here.

On the other hand, I do agree that if we have a `block_time` parameter, we should follow it. If we must follow this, the other solution is to iterate a lot of blocks without updating the light-client to pass over the `trusting_period`. But in my experience, generating tm light clients is expensive, and I couldn't finish the test quickly with the default `trusting_period` - `64000` secs. Maybe we should reduce the `trusting_period` for this test?

File: crates/ibc/src/core/ics02_client/handler/update_client.rs
Comment (rnbguy) [reply crates/ibc/src/core/ics02_client/handler/update_client.rs:257]: looks like `ctx` is the convention for most of the tests. Should I change each of them to `ctx_a`?

Comment (rnbguy) [reply crates/ibc/src/core/ics02_client/handler/update_client.rs:257]: I updated the tests accordingly with `ctx_a`, where `ctx_b` is also used.

File: crates/ibc/Cargo.toml
Comment (rnbguy) [inline crates/ibc/Cargo.toml]: @Farhad-Shabani should we use this to avoid redefining the dev dependencies?

https://github.com/rust-lang/cargo/issues/2911#issuecomment-749580481

With this, we can enable `ibc` crate with `mock` feature and drop the dev-deps for tendermint-testgen, typed-builder and parking-lot.

File: crates/ibc/src/core/ics02_client/handler/update_client.rs
Comment (Farhad-Shabani) [reply crates/ibc/src/core/ics02_client/handler/update_client.rs:257]: For now, I suggest we just keep things right, for this PR. 
Later, we should review all similar misnamings.

File: crates/ibc/src/mock/context.rs
Comment (Farhad-Shabani) [reply crates/ibc/src/mock/context.rs]: > It can happen in real life that the chain was halted but later restarted again from a halted height. In that particular height, the block time can be arbitrarily long. 

This subtle scenario can exist too afaik. Good point of view.
However, given most client expiry would occur during regular operations, I am more inclined towards keeping things consistent the `block_time`, and cutting down the trust period makes more sense to me.

File: crates/ibc/src/mock/context.rs
Comment (Farhad-Shabani) [reply crates/ibc/src/mock/context.rs:222]: On board with this. Should preferably use existing solutions.
Let's discuss this in more detail later on to see how it would be best to derive Builders on different IBC types to keep things consistent and only for testing.

File: crates/ibc/src/mock/host.rs
Comment (Farhad-Shabani) [reply crates/ibc/src/mock/host.rs:42]: > the same was being done in the issue reported by Nomic

I felt why this is added, but sounds like putting these under the `SyntheticTmBlock ` is causing confusion. 

> Actually, I will suggest removing these two fields from SyntheticTmBlock as these two fields are not required for a host context. These fields are only needed when the relayer submits a client update message to the counterparty chain.

Correct. I'll leave it to you. Would be good to come up with better struct definition here. But don't spend too much time on this. Again, as the mock abstraction will be changed in the end..

File: crates/ibc/Cargo.toml
Comment (Farhad-Shabani) [reply crates/ibc/Cargo.toml]: Looks cool. Though not sure how much stable it is. I'll let you decide.

Comment (rnbguy) [reply crates/ibc/Cargo.toml]: I think the workaround is stable and part of the dependency resolver version 2 - which we are using already. But the ideal solution should be more direct.

I will add this, as this only affects dev-dependencies.

Comment (rnbguy) [reply crates/ibc/Cargo.toml]: Just one thing. This will enable `mocks` features by default in the test environment - which is equivalent to the current `cargo test --features mocks` command.

But IMO, this makes sense, as we want to run mock-based tests by default under `cargo test`.

Comment (Farhad-Shabani) [reply crates/ibc/Cargo.toml]: That's fine then.
Just take this into consideration as well: we will soon split out the mock module into a new crate :)

File: crates/ibc/src/mock/host.rs
Comment (rnbguy) [reply crates/ibc/src/mock/host.rs:42]: Let's tackle this in future PR or directly by mock abstraction - as this includes the existing `trusted_height` field.

**Comment by rnbguy** [comment]
@Farhad-Shabani I have made the changes. I also added a new test to check `max_clock_drift`. I will create the issue about the TODO comments after this PR gets merged.

Do I need to add a changelog entry?

**Comment by Farhad-Shabani** [comment]
> @Farhad-Shabani I have made the changes. I also added a new test to check `max_clock_drift`. I will create the issue about the TODO comments after this PR gets merged.

Thanks a lot @rnbguy.

> Do I need to add a changelog entry?

Yes please!

**Review by Farhad-Shabani — APPROVED** [review]
All set 🙏🏻

**Pull request merged**
